### PR TITLE
feat(agent): Add URL whitelisting/blacklisting to config and URL validation

### PR DESCRIPTION
**Pull request by Satyam97**
### Background

This PR refers to the solution for enhancement suggested here: https://github.com/Significant-Gravitas/AutoGPT/issues/5289

This basically takes account white/black listing of URLs during context setting in agents.

### Changes 🏗️

I have added 2 env variables for web policy and URL list. With a combination of both, a URL is white or blacklisted by the read_webpage function.

### PR Quality Scorecard ✨

- [x] Have you used the PR description template? &ensp; `+2 pts`
- [x] Is your pull request atomic, focusing on a single change? &ensp; `+5 pts`
- [x] Have you linked the GitHub issue(s) that this PR addresses? &ensp; `+5 pts`
- [x] Have you documented your changes clearly and comprehensively? &ensp; `+5 pts`
- [x] Have you changed or added a feature? &ensp; `-4 pts`
  - [x ] Have you added/updated corresponding documentation? &ensp; `+4 pts`
  - [ ] Have you added/updated corresponding integration tests? &ensp; `+5 pts`
- [x] Have you changed the behavior of AutoGPT? &ensp; `-5 pts`
  - [ ] Have you also run `agbenchmark` to verify that these changes do not regress performance? &ensp; `+10 pts`

### CODE REVIEW

**Comment by ntindle** [comment]
remove falcon from the PR

File: autogpts/autogpt/.env.template [hunk 1]
```diff
@@ -235,3 +235,12 @@ OPENAI_API_KEY=your-openai-api-key
 # AP_SERVER_PORT=8000
 # # AP_SERVER_DB_URL - Specifies what connection url the agent protocol database will connect to (Default: Internal SQLite)
 # AP_SERVER_DB_URL=sqlite:///data/ap_server.db
+
+
+################################################################################
+### White/Black Listing URLS
+################################################################################
+##WEB_POLICY - Specifies whether the url list is black list(0) or a whilte list(1)
+#WEB_POLICY=0
```
Comment (dboggs95) [inline autogpts/autogpt/.env.template:244]: I think we need a policy for users who want to continue using AutoGPT as normal. This looks like it forces them to choose between whitelisting and blacklisting, which some users might not want either. There needs to be a default option that simply disables this feature.

File: autogpts/autogpt/.env.template [hunk 2]
```diff
@@ -235,3 +235,12 @@ OPENAI_API_KEY=your-openai-api-key
...
+##URL_LIST - Comma separated URLs to be black or whitelisted, sets according to WEB_POLICY value
+#URL_LIST=https://www.google.com,http://www.google.com
```
Comment (dboggs95) [inline autogpts/autogpt/.env.template:246]: The only thing I don't like about putting it here, is that these lists might get pretty long. That's why in my version I was attempting to parse files.

File: arena/falcon.json
Comment (Pwuts) [inline arena/falcon.json:1]: This file doesn't belong in the PR

File: autogpts/autogpt/.env.template (hunk 1 above)
Comment (Pwuts) [inline autogpts/autogpt/.env.template:244]: * Formatting diverges from the rest of the file
* Rather make this an enum with values `blacklist` and `whitelist`

File: autogpts/autogpt/autogpt/config/config.py [hunk 3]
```diff
@@ -241,6 +241,10 @@ class Config(SystemSettings, arbitrary_types_allowed=True):
     # Stable Diffusion
     sd_webui_auth: Optional[str] = UserConfigurable(from_env="SD_WEBUI_AUTH")
 
+    # White/Black Listing URLs
+    web_policy: Optional[str] = os.getenv("WEB_POLICY", 0)
+    url_list: Optional[list] = os.getenv("URL_LIST",[]).split(",")
```
Comment (Pwuts) [inline autogpts/autogpt/autogpt/config/config.py:246]: This is not how the config of the application works, as you can see from how all the other config attributes are defined.

File: autogpts/autogpt/autogpt/url_utils/validators.py [hunk 4]
```diff
@@ -2,10 +2,11 @@
 import re
 from typing import Any, Callable, ParamSpec, TypeVar
 from urllib.parse import urljoin, urlparse
+from autogpt.config import Config
 
 P = ParamSpec("P")
 T = TypeVar("T")
-
+config = Config
```
Comment (Pwuts) [inline autogpts/autogpt/autogpt/url_utils/validators.py:9]: this won't work

File: autogpts/autogpt/autogpt/url_utils/validators.py [hunk 5]
```diff
@@ -2,10 +2,11 @@
 import re
 from typing import Any, Callable, ParamSpec, TypeVar
 from urllib.parse import urljoin, urlparse
+from autogpt.config import Config
```
Comment (Pwuts) [inline autogpts/autogpt/autogpt/url_utils/validators.py:5]: this is gonna throw a linting error

File: autogpts/autogpt/autogpt/url_utils/validators.py [hunk 6]
```diff
@@ -38,6 +41,11 @@ def wrapper(url: str, *args, **kwargs) -> Any:
         # Check URL length
         if len(url) > 2000:
             raise ValueError("URL is too long")
+        if web_policy:
+            if url not in url_list:
+                raise ValueError("URL Not Whitelisted")
+        elif url in url_list:
+            raise ValueError("URL Blacklisted.")
```
Comment (Pwuts) [inline autogpts/autogpt/autogpt/url_utils/validators.py:48]: inconsistent formatting

**Review by Pwuts — CHANGES_REQUESTED** [review]
This still needs some work **and testing** before we can consider merging it

File: autogpts/autogpt/autogpt/url_utils/validators.py (hunk 4 above)
Comment (Satyam97) [reply autogpts/autogpt/autogpt/url_utils/validators.py:9]: @Pwuts can you please help with why this wont work?

File: autogpts/autogpt/autogpt/url_utils/validators.py (hunk 6 above)
Comment (Satyam97) [reply autogpts/autogpt/autogpt/url_utils/validators.py:48]: I have fixed the linting this update the PR with the above config changes @Pwuts

File: autogpts/autogpt/autogpt/url_utils/validators.py (hunk 5 above)
Comment (Satyam97) [reply autogpts/autogpt/autogpt/url_utils/validators.py:5]: Added flake8 to VS code, sorry there was some issue with flake earlier.

File: autogpts/autogpt/autogpt/url_utils/validators.py (hunk 4 above)
Comment (Satyam97) [reply autogpts/autogpt/autogpt/url_utils/validators.py:9]: from autogpt.config import ConfigBuilder
config = ConfigBuilder.build_config_from_env()

web_policy = config.web_policy
url_list = config.url_list

will this work? @Pwuts

**Comment by Swiftyos** [comment]
Closed due to a lack of activity since Feb. If you want to pick this up again let me know and we can reopen
